Add Replicate web demo

CJWBW opened the pull request:
Hey @BreezeWhite! 👋 

This pull request makes it possible to run your model inside a Docker environment, which makes it easier for other people to run it.  We're using an open source tool called [Cog](https://github.com/replicate/cog) to make this process easier.

This also means we can make a web page where other people can try out your model! View it here: https://replicate.com/breezewhite/omnizart

Do [claim your page here](https://replicate.ai/login/github/?next=/breezewhite/omnizart?published=1) so you can own the page, customise the Example gallery as you like, push any future update to the web demo, and we'll feature it on our website and tweet about it too. 

We currently put the model under your GitHub account (main contributor), but we are also planning on supporting GitHub organisation log in and claiming the model.

It is very easy to push update to your model. For faster inference, the checkpoints weights are downloaded to `omnizart/checkpoints`. There is some instruction in `predict.py` if you like to push updated model or weights.

In case you're wondering who I am, I'm from [Replicate](https://replicate.ai/home), where we're trying to make machine learning reproducible. We got frustrated that we couldn't run all the really interesting ML work being done. So, we're going round implementing models we like. 😊

BreezeWhite commented on predict.py:
```diff
@@ -0,0 +1,91 @@
+        help="transcribe mode",
+    )
+    def predict(self, audio, mode):
+        assert str(audio).rsplit('.', maxsplit=1)[-1] == "mp3", "supports mp3 file"
```
Two things to be mentioned.
First, `omnizart` accepts `.wav` files. It seems like the code here are modified from the colab. The reason why colab suggests upload mp3 files is because the upload speed is pretty slow, which takes a pretty long time to upload `.wav` files. So if you want to check the format here, you should also accept `.wav` format.

Second, you could write `str(audio).endswith(".mp3")`, and that is enough. And also the error message should be more verbose. The message can not express clear enough what the actual error is.

BreezeWhite commented on predict.py:
```diff
@@ -0,0 +1,91 @@
+        temp_folder = 'cog_temp'
+        os.makedirs(temp_folder, exist_ok=True)
+        try:
+            audio_name = str(os.path.basename(audio).split(".")[0])
```
It's more safe to re-write the code to `os.path.splitext(os.path.basename(audio))[0]`, since you can't assume there wouldn't be "." in the name.

BreezeWhite commented on predict.py:
```diff
@@ -0,0 +1,91 @@
+            out_path = Path(tempfile.mkdtemp()) / "out.mp3"
+            subprocess.run(["ffmpeg", "-y", "-i", out_name, str(out_path)])
+        finally:
+            clean_folder(temp_folder)
```
Is there any reason to not use `shutil.rmtree(temp_folder)` directly? That should be able to delete everything under the folder.

BreezeWhite commented on predict.py:
```diff
@@ -0,0 +1,91 @@
+        return out_path
+
+
+def clean_folder(folder):
+    for filename in os.listdir(folder):
+        file_path = os.path.join(folder, filename)
+        try:
+            if os.path.isfile(file_path) or os.path.islink(file_path):
+                os.unlink(file_path)
+            elif os.path.isdir(file_path):
+                shutil.rmtree(file_path)
+        except Exception as e:
+            print('Failed to delete %s. Reason: %s' % (file_path, e))
```
Similar to the above comment. If you can use `shutil.rmtree`, then this function is redundant.

BreezeWhite reviewed (changes requested):
Thanks for the contribution very much!! I've left some comments to your changes. Please kindly check them up.
Also there are some missing changes should be done. The documentation page should also be updated since you have added a new badge to the README. More specifically, the [index page](https://github.com/CJWBW/omnizart/blob/cog/docs/source/index.rst) should also be updated with a new short section and description about the playground, below the colab section.

Another question is that as the title mentioned that you added a Docker environment, but there is already Docker image provided of omnizart on Dockerhub. This confused me since, in my opinion, what I see is the new web demo page provided by Replicate, and I don't see where I can download the Docker image you provided. It is also not necessary to provide a new one unless you have made some major changes to omnizart. Maybe change the title to "Add Replicate web demo" is more proper.

CJWBW commented:
Hey @BreezeWhite,

Thank you so much for your support for out work and the through review of the PR! I have updated accordingly. Regarding the docker image, you can find it on the replicate demo page under “Run model with Docker”, but I have renamed the PR to just “Add Replicate web demo” too. Let me know if something still should be changed. Thanks again for the feedback!

BreezeWhite commented on docs/source/index.rst:
```diff
@@ -33,6 +33,11 @@ Colab
 Play with the `Colab notebook <https://bit.ly/omnizart-colab>`_ to transcribe your favorite song almost immediately!
 
+Replicate web demo
+*****
```
The number of "*" should be the same length as line 36.

BreezeWhite commented on docs/source/index.rst:
```diff
@@ -33,6 +33,11 @@ Colab
+Replicate web demo
+*****
+
+Transcribe music with a simple web UI on  `Replicate <https://replicate.com/breezewhite/omnizart>`_.
```
I think this sentence can be rephrased to "Transcribe music with \`Replicate web UI <https://replicate.com/breezewhite/omnizart>\`_."

BreezeWhite commented on scripts/predict.py:
```diff
@@ -5,18 +5,18 @@
 from omnizart.drum import app as dapp
+from omnizart.music import app as mapp
 from omnizart.vocal import app as vapp
 from omnizart.vocal_contour import app as vcapp
```
The modification is on the right path, but still not there. For reference, you can view other files of omnizart the grouping of imports and their order. The first group should be Python built-in libraries, the second should be third-party libraries, and the third should be imports within the project. Between groups should have one empty line to separate them.

BreezeWhite commented on scripts/predict.py:
```diff
@@ -43,13 +43,16 @@ def setup(self):
     def predict(self, audio, mode):
-        assert str(audio).rsplit('.', maxsplit=1)[-1] == "mp3", "supports mp3 file"
-        temp_folder = 'cog_temp'
+        assert str(audio).endswith(".mp3") or str(audio).endswith(".wav"), "please upload mp3 or wav file"
```
The error message should start with upper case, just like normal sentences.

BreezeWhite commented on scripts/predict.py:
```diff
@@ -43,13 +43,16 @@ def setup(self):
         help="transcribe mode",
```
-> "Transcription mode"

BreezeWhite commented on scripts/predict.py:
```diff
@@ -33,7 +33,7 @@ def setup(self):
         "audio",
         type=Path,
-        help="input polyphonic music mp3 file",
+        help="input polyphonic music file, supports mp3 and wav files",
```
Better re-write to "Path to the input music. Supports mp3 and wav format.".

BreezeWhite reviewed (changes requested):
Thanks for the fast response and fixes! There only remains some minor code and sentences need to be modified. After the fixes being committed and pass all the checks, then it's ready for the merge.

CJWBW commented:
Hi @BreezeWhite,

I have modified the code accordingly to your comment! Let me know what you think. Thank you for the helpful feedback!

BreezeWhite reviewed (approved):
LGTM. Thanks for the contribution again ^^
